### ISSUE TITLE
Support `topgrade.toml` inside `topgrade.d/`

### ISSUE THREAD

**Issue by etiennepellegrini** [topgrade-rs/topgrade#624]
I think `topgrade` should check `$XDG_CONFIG_DIR/topgrade.d/topgrade.toml` on top of `$XDG_CONFIG_DIR/topgrade.toml`, and maybe _instead_ of `$XDG_CONFIG_DIR/topgrade/topgrade.toml`.
I don't see the point in having either two topgrade dirs, or a dir and a file in my `$XDG_CONFIG_DIR`

I do not fully understand the use case for `topgrade.d/`, but from the comment in the example config file, it's supposed to contain different config files which are superseded by `topgrade.toml`. I therefore imagine they typically are named differently, and it's not a problem to move `topgrade.toml` to `topgrade.d/topgrade.toml`?

Thanks!

**Comment by niStee** [comment]
@etiennepellegrini Stackexchange provide you with this [answer](https://unix.stackexchange.com/questions/4029/what-does-the-d-stand-for-in-directory-names).

And at least on my amd64 win11, topgrade uses topgrade.toml either in $XDG_CONFIG_DIR/topgrade.d/ $XDG_CONFIG_DIR/topgrade/ or $XDG_CONFIG_DIR/topgrade/.

**Comment by etiennepellegrini** [comment]
Yes, I believe the default should be created in `toprade.d`.
Moreover, the README file should be updated to mention `topgrade.d`, and that the main config file is searched for at `topgrade.d/topgrade.toml` as well.

In addition, `tograde.d/topgrade.toml` should maybe be added explicitly to [config.rs](https://github.com/topgrade-rs/topgrade/blob/788e0412f6ad3e78e040de1cffd84af845550e9e/src/config.rs#L472-L474), as are `tograde/topgrade.toml` and `topgrade.toml`

**Comment by SteveLauC** [comment]
This directory was added before I actually maintained this project, my personal view on it is that we don't need it at all cause Topgrade does not have that many configuration entries and a simple `topgrade.toml` file would be totally sufficient.

**Comment by etiennepellegrini** [comment]
That would seem like the best option to me too 👍

**Comment by SteveLauC** [comment]
Well, that's just my personal point, the previous maintainer thought that this feature should be kept, so I cannot remove it. I would recommend that simply don't use it...

**Comment by bb010g** [comment]
As a side note, as long as we have `topgrade.d`, its semantics should be better defined.

Configuration files in `topgrade.d` are read using [`std::fs::read_dir`]:

<https://github.com/topgrade-rs/topgrade/blob/7c63541cadb93c0370d4ce8ff48604d92bff343c/src/config.rs#L509-L522>

The resulting parsed configurations are then merged together using [`<crate::config::ConfigFile as merge::Merge>::merge>`][`merge::Merge`]:

<https://github.com/topgrade-rs/topgrade/blob/7c63541cadb93c0370d4ce8ff48604d92bff343c/src/config.rs#L543-L560>

The order in which [`std::fs::read_dir`] returns entries is platform and filesystem dependent. That merge operation is not commutative, meaning that the merged configuration will be platform and filesystem dependent when there are 2 or more config files in `topgrade.d`.

My recommendation would be to at least document this, if not also order the configurations using `<std::ffi::OsStr as core::cmp::Ord>` on their file names. [`std::fs::read_dir`] does not guarantee any consistent ordering from consistent platform and filesystem combinations. If we order configurations *and* migrate `${config_dir}/topgrade.toml` into `topgrade.d`, then `${config_dir}/topgrade.toml` should probably become `${config_dir}/topgrade.d/999-user.toml`. Alternatively, our `crate::config::ConfigFile` merge operator can be made commutative, in which case file names can be safely ignored and the user's config can be `topgrade.d/user.toml`, `topgrade.d/topgrade.toml`, or whatever else the user likes. A commutative merge would probably be better long-term, and should be easier to debug.

[`merge::Merge`]: <https://docs.rs/merge/0.1.0/merge/trait.Merge.html>
[`std::fs::read_dir`]: <https://doc.rust-lang.org/std/fs/fn.read_dir.html>

**Comment by GideonBear** [comment]
Decision should be made before v18, so any potential breaking change can be made then

**Comment by MindTooth** [comment]
Frankly, why does even `topgrade.d` exist outside the normal folder?  Ref. [spec](https://specifications.freedesktop.org/basedir/latest/#referencing), `$XDG_CONFIG_HOME/subdir/filename` is correct.  For me it's just clutter that has no purpose other than pollute the config folder with two folder, one being empty because I do not use it (`topgrade.d` in this case).  I fully understand that it might have its use, but read it after someone made it, not force the creation of the folder when it's not in use.

Look at fish.  They use `$XDG_CONFIG_HOME/fish/config.fish`, but also support `$XDG_CONFIG_HOME/fish/conf.d`.  `conf.d` is loaded before `config.fish` (which has bitten me many times. 😭).

I like topgrade as I started to use it more on my systems. 👍🏻   Just annoyed by a folder that is out of place.

**Comment by GideonBear** [comment]
> but read it after someone made it, not force the creation of the folder when it's not in use.

I fully agree with this; if we don't end up changing anything else, at least we should not create it anymore.

**Comment by uwuclxdy** [comment]
Topgrade now doesn't create the `topgrade.d/` folder anymore, still reads the config files from there and merges them like before.

**Comment by uwuclxdy** [comment]
My thoughts on this: Lets just **deprecate `topgrade.d/`** in favor of `[include]`.

`[include]` already does the same job but explicitly: files are ordered by the user and are on the same parse path as the main config. `topgrade.d/` is a separate code path, which is also where #852 (symlinks ignored) and #1222 (tilde not expanded there) came from.

I'd suggest:
- pre-v18: **warn** if `topgrade.d/` contains files, with a migration hint.
- v18: drop support entirely. It should hard-error with migration hint if files are still there rather than silently drop config, since ignoring a `disable` list could run steps someone turned off in that config for example.

**Comment by GideonBear** [comment]
I disagree; I think we should keep `topgrade.d` supported and documented, *but merge the codepath with the includes*. There's a legitimate use for `topgrade.d`, and it isn't in the way for anyone (anymore, now that it's not automatically created anymore 😄).
